File: scenebuilder/src/main/java/com/sfxcode/sapphire/javafx/skin/SFXDataListViewSkin.java
package com.sfxcode.sapphire.javafx.skin;

import com.sfxcode.sapphire.javafx.control.SFXDataListView;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.control.Label;
import javafx.scene.control.ListView;
import javafx.scene.control.SkinBase;
import javafx.scene.control.TextField;
import javafx.scene.layout.VBox;

import java.util.ArrayList;
import java.util.List;


public class SFXDataListViewSkin<S> extends SkinBase<SFXDataListView<S>>{

    public VBox box = new VBox(5);
    public TextField textField = new TextField("Search Field");
    public Label footerLabel = new Label("DataListView Footer");

    private List<String> list = new ArrayList<String>();
    private ObservableList<String> observableList = FXCollections.observableList(list);
    public ListView<String> listView = new ListView<String>(observableList);

    /**
     * Constructor for all SkinBase instances.
     *
     * @param control The control for which this Skin should attach to.
     */
    public SFXDataListViewSkin(SFXDataListView<S> control) {
        super(control);
        observableList.add("Item 1");
        observableList.add("Item 2");
        observableList.add("Item 3");
        updateView();
    }

    public void updateView() {
        getChildren().add(box);
        box.getChildren().add(textField);
        box.getChildren().add(listView);
        box.getChildren().add(footerLabel);
    }

}

File: src/main/scala/com/sfxcode/sapphire/javafx/skin/SFXDataListViewSkin.scala
package com.sfxcode.sapphire.javafx.skin

import javafx.scene.control.SkinBase
import com.sfxcode.sapphire.javafx.control.{ SFXDataListView, SFXListCellFactory }
import javafx.scene.control._
import javafx.scene.layout.{ HBox, VBox }
class SFXDataListViewSkin[S <: AnyRef](view: SFXDataListView[S]) extends SkinBase[SFXDataListView[S]](view) {

  val contentBox = new VBox()
  contentBox.setSpacing(5)
  contentBox.getStyleClass.add("sfx-data-list-content-box")

  val label = new Label("Footer Label")
  label.getStyleClass.add("footer-label")
  view.footerLabel.set(label)

  val headerBox = new HBox
  headerBox.getStyleClass.add("sfx-data-list-header-box")
  view.header.set(headerBox)

  val footerBox = new HBox
  footerBox.getStyleClass.add("sfx-data-list-footer-box")
  footerBox.getChildren.add(label)
  view.footer.set(footerBox)

  updateCellFactory()
  view.cellProperty.addListener((_, _, _) => updateView())

  view.header.addListener((_, _, _) => updateView())

  view.showHeader.addListener((_, _, _) => updateView())

  view.footer.addListener((_, _, _) => updateView())
  view.showFooter.addListener((_, _, _) => updateView())

  getChildren.add(contentBox)

  updateView()

  def updateView(): Unit = {
    contentBox.getChildren.clear()
    if (view.showHeader.get && view.header.get != null)
      contentBox.getChildren.add(view.header.get)

    contentBox.getChildren.add(view.listView)

    if (view.showFooter.get && view.footer.get != null)
      contentBox.getChildren.add(view.footer.get)
  }

  def updateCellFactory(): Unit = {
    val cellFactory = new SFXListCellFactory[S]
    cellFactory.setProperty(view.cellProperty.get)
    view.listView.setCellFactory(cellFactory)
  }

}
